src/lib/bmp/bmp.c

```diff
@@ -0,0 +1,244 @@
+
+	read(fd,&(bmp_head->bfSize), 4);
+
+	read(fd,&(bmp_head->bfReserved), 4);
+
+	read(fd,&(bmp_head->bfOffBits), 4);
+
+	read(fd,&(bmp_head->biSize), 4);
+
+	read(fd,&(bmp_head->biWidth), 4);
+
+	read(fd,&(bmp_head->biHeight), 4);
+
+	read(fd,&(bmp_head->biPlanes), 2);
+
+	read(fd,&(bmp_head->biBitCount), 2);
+
+	read(fd,&(bmp_head->biCompression), 4);
+
+	read(fd,&(bmp_head->biSizeImage), 4);
+
+	read(fd,&(bmp_head->biXPelsPerMeter), 4);
+
+	read(fd,&(bmp_head->biYPelsPerMeter), 4);
+
+	read(fd,&(bmp_head->biClrUsed), 4);
+
+	read(fd,&(bmp_head->biClrImportant), 4);
+
+	return 0;
+}
+
+static void read_pallete(int fd, struct color pallete[], int colors) {
+	uint8_t tmp[4];
+	int i;
+
+	lseek(fd, HEADER_END, SEEK_SET);
+	for (i = 0; i < colors; i++) {
+		read(fd, tmp, 4);
+		pallete[i].red = tmp[1];
+		pallete[i].green = tmp[2];
+		pallete[i].blue = tmp[3];
+	}
+}
+
+int get_image_bmp(int fd, struct bmp_header *bmp_head, uint8_t *image) {
+	int i;
+	uint8_t pixel;
+	struct color pallete[256];
+
+	if (check_bmp(fd) != 0) {
+		return -1;
+	}
+
+	get_head_bmp(fd, bmp_head);
+	switch(bmp_head->biBitCount) {
+		case 32:
+			lseek(fd, bmp_head->bfOffBits, SEEK_SET);
+			read(fd, image, bmp_head->biWidth * bmp_head->biHeight * 4);
+			break;
+
+		case 24:
+			lseek(fd, bmp_head->bfOffBits, SEEK_SET);
+			read(fd, image, bmp_head->biWidth * bmp_head->biHeight * 3);
+			break;
+
+		case 8:
+			read_pallete(fd, pallete, (bmp_head->bfOffBits - 54) / 4);
+			lseek(fd, bmp_head->bfOffBits, SEEK_SET);
+			for (i = 0; i < bmp_head->biWidth * bmp_head->biHeight * 3; i += 3) {
+				read(fd, &pixel, 1);
+				image[i] = pallete[pixel].blue;
+				image[i + 1] = pallete[pixel].green;
+				image[i + 2] = pallete[pixel].red;
+			}
+			break;
+
+		default:
+			printf("invalid bitcount\n");
+			return -1;
+	}
+	return 0;
+}
+
+static int get_fbp_bmp(struct fb_var_screeninfo *vinfo, char *buf, uint8_t *img,
+									uint32_t width, uint32_t height, uint32_t bits) {
+	int i, j, offset, addres, x = 0;
+	int fb_bytes_per_pixel = vinfo->bits_per_pixel / 8;
+
+	for(i = height; i > 0; i--) {
+		offset = 0;
+		for(j = width; j > 0; j--) {
+			addres = offset + i * vinfo->xres * fb_bytes_per_pixel;
+			offset += fb_bytes_per_pixel;
+
+			if (vinfo->bits_per_pixel == 32) {
+				* (buf + addres) = img[x];
+				* (buf + addres + 1) = img[x + 1];
+				* (buf + addres + 2) = img[x + 2];
+				* (buf + addres + 3) = 255;
+				if (bits == 32) {
+					x += 4;
+				} else {
+					x += 3;
+				}
+			}
+		}
+	}
+	return 0;
+}
+
+int draw_fb_bmp(char *file_name) {
+	int img_size, fd = 0, fbfd = 0;
+	struct fb_var_screeninfo vinfo;
+	struct fb_fix_screeninfo finfo;
+	struct bmp_header bmp_head;
+	long int screensize = 0;
+	char *fbp = 0;
+
+	fd = open(file_name, O_RDONLY);
+	if (fd == -1) {
+		printf("Error : can't open file\n");
+		return -1;
+	}
+
+	if (check_bmp(fd) != 0) {
+		return -1;
+	}
+
+	get_head_bmp(fd, &bmp_head);
+
+	if (bmp_head.biBitCount == 32) {
```

Comment:
yes, it can be 32(RGB + alpha),  i reworked this place and now not adding alpha in image buffer, also i added support of 1, 4 bpp and fix bugs with 24 bpp, now its support all common formats of bmp

Reply:
the size of the picture will always be bmp_head.biWidth * bmp_head.biHeight * 3, because regardless of the BMP version, each pixel is given by three numbers